proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1284 @@
+Compilation does not occur unless *all* required signatures are implemented,
+allowing a compiler can optimize as if Backpack was not present (*no
+performance overhead.*)  An indefinite library can also be partially
+instantiated, or not instantiated at all.  If a required signature is not
+instantiated, it gets inherited by the user of the library::
+
+    -- in the Cabal file
+    library stringutils-indef
+        -- No Str module in scope, so Str is left uninstantiated,
+        -- giving stringutils-indef an (implicit) requirement
+        -- on Str.
+        build-depends: concat-indef
+        exposed-modules: StringUtils
+
+    -- in StringUtils.hs
+    module StringUtils where
+    import Concat
+    import Str -- the signature is importable
+
+It's worth reiterating that contents of a ``signatures`` field
+do not specify the *required* signatures of a library, since a library
+may also inherit many other required signatures from its dependencies.
+(TODO: A user can explicitly specify all implicit signatures using
+the ``implicit-signatures`` field.)
+
+Backpack is quite flexible about the way the uninstantiated
+required signatures can be handled:
+
+* If you depend on two indefinite libraries, both of which
+  have the same required signature (e.g., ``Str``), then you
+  have a single required signature ``Str`` that is the union
+  of these two signatures: signatures are identified only
+  by module name.  To keep these two requirements separate,
+  you would rename one of the requirements to a different name
+  using the ``backpack-includes`` directive::
+
+    library one-string
+        -- One requirement, named Str
+        build-depends: concat-indef, stringutils-indef
+
+    library two-string
+        -- Two requirements, Str and Str2
+        build-depends: concat-indef, stringutils-indef
+        backpack-include:
+            stringutils-indef requires (Str as Str2)
+
+* In addition to the inherited requirements from dependencies,
+  a user can also define a local ``hsig`` to refine the required
+  signature further (i.e., define extra types).
+
+The current implementation of Backpack in GHC has some notable
+user-facing limitations:
+
+1. It is not possible to define a module in a library, and then
+   use it to immediately to instantiate an indefinite library::
+
+        library concat-bytestring-bad
+            build-depends: concat-indef -- has Str requirement
+            exposed-modules: Str, ConcatUser -- can't use these to fill
+
+   Instead, ``Str`` must be pulled out into a separate library
+   of its own (Cabal 2.0 supports multiple libraries in a package,
+   making this less burdensome.)  The reason for this restriction
+   is to simplify implementation of the build system: if this
+   mode of use was allowed, it would be necesary to first build
+   ``Str``, then build ``concat-indef``, and then come back to
+   ``concat-bytestring-bad`` and finish building the rest of the
+   modules.
```

Comment:
Good; I remember this being a stumbling block in previous discussions.  I like the approach of multiple libraries per package to reduce the administrative overhead.